### ISSUE TITLE
newrelic_rpm 6.13.0 causes elasticsearch results to fail

### ISSUE THREAD

**Issue by sighmon** [newrelic/newrelic-ruby-agent#450]
Version `6.13.0` seems to be incompatible with elasticsearch-rails `7.1.1`

## Description
When updating from `6.12.0.367` to `6.13.0` the search results from elasticsearch only returned the first ten results, and the pagination failed.

## Expected Behavior
Elasticsearch returns full results, and pagination works.

## [Troubleshooting](https://discuss.newrelic.com/t/troubleshooting-frameworks/108787) or [NR Diag](https://docs.newrelic.com/docs/using-new-relic/cross-product-functions/troubleshooting/new-relic-diagnostics) results
Unfortunately there weren't any errors. It silently returned the results mentioned above.

## Steps to Reproduce
Unfortunately I don't have time to reproduce this bug in a repository, sorry!

## Your Environment
Heroku
Bonsai Elasticsearch
ruby 2.7.1
rails 5.2.4.4
elasticsearch-rails 7.1.1
newrelic_rpm 6.13.0

## Additional context
Trying to view NewRelic data from Heroku's dashboard is also broken. I can't see any error logs anymore.

**Comment by tannalynn** [comment]
Thank you for bringing this issue to our attention. We will look into this.

**Comment by mwlang** [comment]
@sighmon Thank you for opening this issue.  Hopefully we can get to the bottom of this one, but there's not a lot of information to go on.  

First, I am assuming that reverting back to `6.12.0.367` resolves this issue?  If not, then this suggests another change outside the agent update is causing this issue.  If using 6.12 does resolve current issue, then can you share what instrumentation code (such as `add_method_tracer`, `add_custom_attributes`, etc.) might be wrapping your calls to ES?  

In purely speculative terms, there were two changes that _might_ have an effect on ES results returned in 6.13:

* calling `add_custom_attributes` no longer modifies the params of the caller

  The previous agent's improvements to recording attributes at the span level had an unexpected side-effect of modifying the params passed to the API call as duplicated attributes were deleted in the process. This is now fixed and params passed in are no longer modified.

* `http.url` query parameters spans are now obfuscated

  Previously, the agent was recording the full URL of the external requests, including the query and fragment parts of the URL as part of the attributes on the external request span. This has been fixed so that the URL is obfuscated to filter out potentially sensitive data.

All other changes were relatively minor and do not seem to code that might play a role in this change in behavior.   Does either of the above make sense in the context of how your application might be integrating the agent around ES queries?

**Comment by marcosdemelo** [comment]
I also found error. When upgrading from version 6.12.0.367 to 6.13.0, we also faced serious problems with the [Savon](https://github.com/savonrb/savon) gem, which we used in the project for SOAP communication.

We reverted the version and everything went back to normal.

In the next version, I will test the site before putting it into production. Because the whole test suite passed without problems. The error happened only in Rails 6 production mode for me.

**Comment by jduff** [comment]
@mwlang I'm seeing the same issue in a Rails 6 application talking to elastic search as well. We're not doing anything with `add_method_tracer` or `add_custom_attributes` in our application and I didn't see any specific references to NewRelic in the gems we are using either. Rolling back to `6.12.0.367` fixed the issue for me as well.

I looked through the diff and couldn't find anything obvious that could be causing this either. I tested it development, enabling new relic monitoring for my development environment and didn't see the same issue, is there an easy way to put the agent into 'production' mode to test it in development to see if I can reproduce the error there?

If there's anything else I can do to debug this in development or more information that would be helpful please let me know!

**Comment by sighmon** [comment]
@mwlang Yes, reverting to `6.12.0.367` solved the issue for me. I couldn't replicate it in Development either, as @jduff noted above.

Here's the code I've been using on my magazine `Issue` model:

```ruby
  def self.search(params, admin = false)
    pagination = Settings.issue_pagination
    clean_query = params[:query].try(:gsub, /[^0-9a-z "]/i, '')
    if admin
      pagination = 200
    end
    search_hash = {
      sort: [{ release: {order: "desc", "unmapped_type": "long"}}]
    }
    search_hash.merge!({query: { query_string: { query: clean_query, default_operator: "AND" }}}) if params[:query].present?
    search_hash.merge!({ post_filter: { term: { published: true}} }) unless admin

    __elasticsearch__.search(search_hash).page(params[:page]).per(pagination).records
  end
```

**Comment by mwlang** [comment]
Highly likely this is resolved by resolving #451

**Comment by mwlang** [comment]
@sighmon I believe I have a potential fix for this based on the issue reported in #451.  I do not have an ES reproduction setup to confirm this fixes your issue.  Would you be willing to test the potential fix out?  If so, modify your Gemfile `newrelic_rpm` entry to install from the github branch where the fix resides as follows:

```ruby
gem 'newrelic_rpm', 
  git: 'https://github.com/newrelic/newrelic-ruby-agent.git', 
  branch: 'query_params_lost_for_external_requests'
```
The version number will not change, but you should see the branch and commit hash reported in the `bundle install` output: 
```
Using newrelic_rpm 6.13.0 from https://github.com/newrelic/newrelic-ruby-agent.git (at query_params_lost_for_external_requests@43501ed)
```

**Comment by jduff** [comment]
@mwlang I tested this in my staging environment, first with released 6.13.0 to see it break, then with the `query_params_lost_for_external_requests` branch to see it fixed and am not seeing the issue with the fixed branch.

Thanks for digging into this so quickly!

**Comment by mwlang** [comment]
@jduff thank you for confirming this fixes the issues you saw.  Generally speaking, the agent works the same in production as in development, so it's surprising it didn't replicate in both environments.  One reason it might not have replicated is because the bug depends on a `::URI` being passed in vs. just a `String` representation of a URL.  Our recent improvement to obfuscate URLs was passing that original `::URI` along and modifying it when it should not have.

**Comment by basex** [comment]
Would be good if you could release a version with the fix, this also broke our SOAP services in production and it took us 1 hour to identify the root cause was this gem.

**Comment by mwlang** [comment]
@basex Thank you for weighing in.  We were internally debating whether to push this now or later with a regular release cycle.  We're going to go ahead and push a patch release.

**Comment by tannalynn** [comment]
The fix for this issue has been released today in version 6.13.1. Thank you for bringing it to our attention.

**Comment by sighmon** [comment]
@mwlang I can confirm that `6.13.1` has fixed this issue for me. Thank you.

**Comment by angelatan2** [comment]
Hello @jduff 

> @mwlang I tested this in my staging environment, first with released 6.13.0 to see it break, then with the `query_params_lost_for_external_requests` branch to see it fixed, and am not seeing the issue with the fixed branch.
> 

Thanks for helping us test this in your staging environment.  We really appreciate it. The maintainers of this repo have recommended your work.  If you'd like us to send you a swag as a token of our gratitude, please write to angelatan@newrelic.com.  Otherwise,  thank you again!
